fix: fake-lmp-device-register to test the solution locally

camilamacedo86 opened the pull request:
**Description**

Fix the fake client to emulate the device registration and generate the /var/sota/ dir with the files in the current directory when we are using it to do the tests.

See:

```
$ python3 fake-lmp-device-register --registration-url "http://localhost:80/sign" --factory fioed-camila-sample

2023-11-02 09:10:17,901 - __main__ - INFO - Starting main function with options: Factory=fioed-camila-sample, SOTA Directory=/Users/camilamacedo/go/src/github.com/foundriesio/factory-registration-ref/var/sota
2023-11-02 09:10:17,901 - __main__ - INFO - Creating key for UUID: dc4d274b-80ce-471b-87a2-a50a31b35fca, Factory: fioed-camila-sample
2023-11-02 09:10:17,916 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): localhost:80
2023-11-02 09:10:18,185 - urllib3.connectionpool - DEBUG - http://localhost:80 "POST /sign HTTP/1.1" 201 4057
2023-11-02 09:10:18,185 - __main__ - DEBUG - Successfully registered device with status code: 201
```

Signed-off-by: Camila Macedo <camila.macedo@foundries.io>

doanac commented on fake-lmp-device-register:
```diff
@@ -50,19 +56,35 @@ extendedKeyUsage=critical, clientAuth
+            os.unlink(keyfile_path)
+            raise subprocess.CalledProcessError(r.returncode, r.args)
+        os.unlink(cnf.name)
+        os.unlink(keyfile_path)
```
looks like you are making two changes in one commit and I'm not sure I understand what keeping these temp files around is actually achieving?

vkhoroz replied:
@doanac you beat me to it :-)

@camilamacedo86 could you either split this PR into several commits, or only leave what is relevant for the fix itself.
It is hard to grasp what was the problem and what is the fix.

camilamacedo86 replied:
If we break each one of them will not work alone 
But I will try to do so

doanac replied:
its not so much that - it's why did you have to change this? I think this part of the code worked and was simpler to follow.

and then there's logging changes and a change to the default location it writes to.  So what is broke and what needs to change.

camilamacedo86 replied:
This part did not work. The error was faced when we were trying to read the key from the input.
I tried other ways, but the way that I could make it work was ensuring a temp file to be used.

Error we fixed (Example):
    
    $ python3 fake-lmp-device-register --registration-url "http://localhost:80/sign" --factory fioed-camila-sample private key from -: No such file or directoryTraceback (most recent call last):  File "/Users/camilamacedo/go/src/github.com/foundriesio/factory-registration-ref/fake-lmp-device-register", line 146, in <module>    main(options)  File "/Users/camilamacedo/go/src/github.com/foundriesio/factory-registration-ref/fake-lmp-device-register", line 65, in main    pkey, csr = create_key(opts.uuid, opts.factory, opts.production)                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  File "/Users/camilamacedo/go/src/github.com/foundriesio/factory-registration-ref/fake-lmp-device-register", line 59, in create_key    r.check_returncode()  File "/opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 502, in check_returncode    raise CalledProcessError(self.returncode, self.args, self.stdout,subprocess.CalledProcessError: Command '['openssl', 'req', '-new', '-config', '/var/folders/zr/51746skj0gz6qnt27hmvvbnc0000gn/T/tmp5ph4z49y', '-key', '-']' returned non-zero exit status 1.

camilamacedo86 replied:
But I will add a try catch to ensure that we delete always the temp files. 
hold on I am pushing a PR with the commits and some nit improvements

camilamacedo86 commented:
@doanac @vkhoroz 

Thank you for your attention and time.
See that I split the changes with the commits
The error faces to clarify it are also added in the commits messages/logs

I hope that it makes the motivations more explicit.
Please, feel free to let me know wdyt.

doanac commented on fake-lmp-device-register:
```diff
@@ -29,43 +29,62 @@ class Options(NamedTuple):
-        cnf.flush()
-
-        r = subprocess.run(
-            ["openssl", "req", "-new", "-config", cnf.name, "-key", "-"],
```
Now I see openssl broke something on us. I can confirm this no longer works under Arch Linux either.

camilamacedo86 replied:
We removed this one.
However, please let us know if you see that is broken in the linux

doanac commented on fake-lmp-device-register:
```diff
@@ -29,43 +29,62 @@ class Options(NamedTuple):
+    finally:
+        # Let's ensure that temp files are deleted
+        os.unlink(cnf.name)
+        os.unlink(keyfile_path)
```
I think you can simplify this quite a bit by using both tempfile in the same context. e.g:
```
keyfile = NamedTemporaryFile()
cnf = NamedTemporaryFile()
with keyfile, cnf:
    # key file logic
    keyfile.flush()
    # create the csr
```

vkhoroz replied:
Just IMHO: we don't need the key file to be a temp file, because the next thing we do at line 97 is... save the key file to `opts.sota_dir`.

So, I'd restructure this code to do the following:
1. Save the cnf into a tmp file.
2. Generate the key file into `{opts.sota_dir}/pkey.pem`.
3. Generate csr into stdout (as is now) and return it.

That should save a lot of tmpfile and file.flush operations.
Moreover, it would allow to keep the original code structure to the max, as we'd only need to change two operations - to generate the key, and to use the key.

camilamacedo86 replied:
+1 The suggestion was applied

doanac commented on fake-lmp-device-register:
```diff
@@ -23,46 +28,77 @@ class Options(NamedTuple):
+        raise
+    finally:
+        # Let's ensure that temp files are deleted
+        os.unlink(cnf.name)
```
you don't have to do these unlink calls. let the context delete the tempfile by removing `delete=False`, and remove the call to `cnf.close()` as the `cnf.flush` is enough to get it to disk.

camilamacedo86 commented:
Hi @doanac and @vkhoroz 

I addressed the suggestions and broke the changes in commits and other PR's as suggested by @vkhoroz 
I hope that helps us to move forward easily. So see:

- https://github.com/foundriesio/factory-registration-ref/pull/18
- https://github.com/foundriesio/factory-registration-ref/pull/19
- https://github.com/foundriesio/factory-registration-ref/pull/20
